django/contrib/contenttypes/models.py

```diff
@@ -96,12 +96,20 @@ def get_for_models(self, *models, for_concrete_models=True):
                     results[model] = ct
                 self._add_to_cache(self.db, ct)
         # Create content types that weren't in the cache or DB.
-        for opts, opts_models in needed_opts.items():
-            ct = self.create(
-                app_label=opts.app_label,
-                model=opts.model_name,
-            )
```

Comment:
We have to remember that `bulk_create()` doesn't call `save()` and will not emit signals, so this change can be backward incompatible for some projects.

Also, `bulk_create()` will not return new content types with primary keys on some databases, e.g. MySQL, Oracle, MariaDB < 10.5, or SQLite < 3.35.

Not sure it's worth changing.

Reply:
Ugh, yeah, that’s true. And indeed, in nearly all cases, the content types are created by `django.contrib.contenttypes.management.create_contenttypes()`, which does use `bulk_create()`. If I undo this `bulk_create()` change, my profiling remains the same, since this creation section wasn’t being called. So I’ll revert changes to `get_for_models`.